Sk runtime effect binding

HinTak opened the pull request:
Beginning of addressing #269. Include #272 and #267 , but not #273 (that needs tidying).

Runs https://github.com/HinTak/skia-python-examples/blob/main/SkSL_MinifiedSkSL.py . Should run 2 of the 9 examples now.

HinTak commented:
All the c++ examples from upstream skia has been ported to python : the `SkSL_* .py` ones https://github.com/HinTak/skia-python-examples/

HinTak commented:
@jonathanhogg @kyamagu fedora 41 released a few days ago ships python 3.13, so we basically want 3.13 wheels for the next release, really. I have pull #277 & #275 in. If m132 is not much of a jump, will have this merge into that quickly just before m132 goes out. Still missing are some tests & docs (that will be README.m132, hopefully).

HinTak commented:
@kyamagu this is a fairly large set of changes so I thought I'd ask for review early. Ready to go I think, except I'd like to pull in skia m132 when it comes out in a week or so. That means the usual version bump, plus adding whatever is new on top of README.m132. I'll do a separate new pull based on #275, and pull it when that happens. Other than that, README.m131 contains basically a description of #275, while README.m132 contains everything else, plus something to be added when upstream m132 happens. Read those two files first .

kyamagu commented on src/skia/RuntimeEffect.cpp:
```diff
@@ -0,0 +1,197 @@
+
+py::class_<SkRuntimeEffectBuilder> runtime_effect_builder(m, "RuntimeEffectBuilder");
+
+py::class_<SkV3>(m, "V3")
```
Is supporting an implicit conversion from `list` or `tuple` helpful for users, like `SkColor4f`?

HinTak replied:
Yes, I think v4 those are almost like color4f's except it can take half-precisions. It is all new to but I can check

HinTak replied:
I added implicit conversion for `tuple` and thought it didn't work, so I overloaded `setUniform` to take `tuple` and found I was passing a `list` (always a bit confused about which is which myself), I like the list syntax a bit better; but now both works, just via different mechanisms!

Am a bit curious how python chooses, if tuple is implicit convertible to both V3 and V4 (depending on length = 3 or 4).

kyamagu commented on tests/test_font.py:
```diff
@@ -643,3 +658,47 @@ def test_FontMetrics_hasStrikeoutThickness(fontmetrics):
+@pytest.fixture
+def color_emoji_run():
+    if sys.platform.startswith("linux"):
+        if (os.getenv("GITHUB_ACTION") == True):
```
Maybe check the file with `os.exists()` instead of relying on the CI runner environment?

HinTak replied:
That's a good idea. I'll do that. Strangely the font is available on Ubuntu but not accessible via fontconfig. (And at a different place on fedora)

HinTak replied:
Done that.

HinTak commented on relnotes/README.m132.md:
```diff
@@ -0,0 +1,24 @@
+  `SkRuntimeEffect::Result` as `RuntimeEffectResult`, 
+  `SkRuntimeEffect::ChildPtr` as `RuntimeEffectChildPtr`,
+  `SkRuntimeEffectBuilder` as `RuntimeEffectBuilder',
+  `std::vector<SkRuntimeEffect::ChildPtr>` as `VectorSkRuntimeEffectChildPtr`,
```
Yes, removing it would be better. I'll do that.

HinTak commented:
Ubuntu x86_64 python 3.13:
 2147 passed, 102 skipped, 18 xfailed, 27 warnings

Passed gone up a fair amount since m130 because of the new runtime effect tests; xfail up slightly from withdrawal of 3 patheffect.* in m132.

@kyamagu - there are not many interesting changes since you looked at it - m132 (withdrawal of 3 small api from public), and I added a few more overloads to runtime effects and also updated / added some tests. The slightly confusing part is the test changes: there are 3 ways of making runtime effects (for the purpose of making shaders, color filters, and blenders). When you looked at it, I only had tests and examples for the first one. We can't quite provide the last one as we don't even bind `SkBlender` yet. So the tests are updated for the 2nd case (and the minimal sksl snipplet for each of the three) and the beginning of re-using the created color filter for more interesting usages and tests.

HinTak commented:
Mac os Intel python 3.13, for comparing with m130:
2140 passed, 110 skipped, 18 xfailed, 27 warnings
(I didn't mention ubuntu in m130 due to #266, which is now fixed)
